### PR TITLE
Add the option to exclude certain layers from the tarball.

### PR DESCRIPTION
**Pull request by mattmoor**
This should be considered a relatively advanced option, but for folks that know
what they are doing you can reduce the amount of data that you need to encode in
the tarball for the daemon to load it.

The ultimate use case of this option will be from `daemon.Write`, which
currently uses the `docker load` interface to pull image into the daemon,
however, this currently reuploads (and redownloads) the base image on each write
in context like `ko`. If we can determine the set of layers that already exist
in the daemon we can elide these from the tarball to dramatically improve
performance.

Related: #205

### CODE REVIEW

File: pkg/v1/tarball/write.go [hunk 1]
```diff
@@ -89,6 +105,15 @@ func Write(tag name.Tag, img v1.Image, wo *WriteOptions, w io.Writer) error {
 		// https://www.gnu.org/software/gzip/manual/html_node/Overview.html
 		layerFiles[i] = fmt.Sprintf("%s.tar.gz", hex)
 
+		// We filter late because the lenght of layerFiles must match the diff_ids
```
Comment (jonjohnsonjr) [inline pkg/v1/tarball/write.go]: nit: `length`

File: pkg/v1/tarball/write.go [hunk 2]
```diff
@@ -66,6 +72,16 @@ func Write(tag name.Tag, img v1.Image, wo *WriteOptions, w io.Writer) error {
 		return err
 	}
 
+	// Default filter function
```
Comment (jonjohnsonjr) [inline pkg/v1/tarball/write.go]: Maybe hoist this out into its own method?

File: pkg/v1/tarball/write.go [hunk 3]
```diff
@@ -66,6 +72,16 @@ func Write(tag name.Tag, img v1.Image, wo *WriteOptions, w io.Writer) error {
...
+	filter := func(_ v1.Layer) (bool, error) {
+		return true, nil
+	}
+	if wo != nil {
+		if wo.LayerFilter != nil {
```
Comment (jonjohnsonjr) [inline pkg/v1/tarball/write.go]: Collapse this into `if wo != nil && wo.LayerFilter != nil {` ?

Though I guess that makes it harder to add more options... up to you.

**Comment by mattmoor** [comment]
@jonjohnsonjr I hear you, but at the same time, I'm not sure of a good way to do this closer to `daemon.Write` that doesn't also involve cloning a significant portion of `tarball.Write`.

**Comment by dlorenc** [comment]
can you rebase this?

**Comment by mattmoor** [comment]
At this point, it's more work to rebase than redo.  @jonjohnsonjr are you up for it?

**Pull request closed without merging**